### PR TITLE
Add packet inspection hook methods

### PR DESCRIPTION
**Pull request by pschou**

### CODE REVIEW

**Comment by TimRots** [comment]
Hi @pschou. Thank you for the PR!
Can you provide some more information on why this features is wanted, how it's used, etc.

@SuperQ can you share some thoughts on this?

**Comment by pschou** [comment]
@TimRots I wanted the ability to know the network latency without having to do a separate ping.  The problem with wrapping the Get / GetBulk with a timing clause is that when the query is sent again for a retry-- the actual latency is not measurable.  So pushing this into the Get / GetBulk function will ensure that the real latency is measured and pushed back out with the SNMP result.

**Comment by gaissmai** [comment]
@pschou this latency is different to the icmp/ping latency, yes I'm sure you know.
And yes, for me it would be also useful to get the agents SNMP latency with your code addition, a lot of them have an awful SNMP implementation, with this latency I could sort the slow agents to a different poll queue, just my 5 Cents

**Comment by SuperQ** [comment]
Interesting idea. Typically I wrap calls on the outside, but sometimes it is interesting to get more granular data.

One interesting idea is we could add an optional way to register an instrumentation function. This way packet responses could be sent to an `Observe()` method.

**Comment by gaissmai** [comment]
@SuperQ sending the **response** would not be enough to measure the latency, you are missing the reference timestamp when you sent the request. You would also have to send the **request** to your observer. Maybe this generalized API would be overkill and not KISS for this low level SNMP library.

**Comment by pschou** [comment]
@gaissmai Yup!  This difference between vendors is noticeable and I purposefully send a very short Get to simulate a roundtrip time and then run subsequent requests and measure the spread of the packets.  A fun trick-- kind of like a controlled amplification backscatter measurement-- with multiple UDP packets coming back, I can measure (to a degree) the spacing between the packets and infer the available link bandwidth.  A neat metric to watch, but not qualitative at all-- :)

@SuperQ I like to keep it simple.  So the Observe method would be interesting...  The packet-count would be nice too (for what is stated above), but implementation is key...  each call is taking up clock cycles, and with operating at the millisecond level I don't feel comfortable adding extra calls between the write and read.  Just my 10 cents.

**Comment by SuperQ** [comment]
What I'm thinking about is an "on receive" function. Similar to how it's handled in this [ping library](https://pkg.go.dev/github.com/go-ping/ping#Pinger). We still add the packet duration like we're doing here.

I use something like this for both logging and monitoring in [this Prometheus prober utility](https://github.com/SuperQ/smokeping_prober/blob/9ba85274dcc21bf8132cbe3b3dccfcb4aab57d9f/collector.go#L66-L71).

I could see this being useful for general observability of this library, without tying it to a specific metrics implementation. Although, I'd not object to Prometheus being included in gosnmp. :grinning:

**Comment by pschou** [comment]
@SuperQ Prometheus is my goal indeed!!!  With the recent SolarWinds development... a selfish plug here... I've rewritten the guts of the snmp monitoring that SolarWinds provided using prometheus and grafana here.  Check it out as I literally just punched it out this weekend:

https://github.com/pschou/snmp-prom

**Comment by SuperQ** [comment]
Have you seen the [official snmp_exporter](https://github.com/prometheus/snmp_exporter)?

**Comment by pschou** [comment]
Yup, didn't like the implementation at all.  The problem is that when I am evaluating links and need to sync up data on the packet counts/errors, the time that the scrape is done is *very* important to capture the time windows and compare both sides at the same time.  The package I wrote takes this into account and does a "time to run" - latency/2 so as to get the metrics on all the routers/switches to sync up to the sub second.  So getting the latency really matters.  ;)

**Comment by pschou** [comment]
I should specify, I am doing interface In - interface Out subtraction on both sides of the link...

**Comment by pschou** [comment]
@SuperQ (and @gaissmai) Upon further thought into the Observer() method, I realize that am in error and I concede that you are indeed right that an observer handle which could be turned on and off by setting a handle would give hooks for forward flexibility.  I am imagining a sleep statement that holds a query write until released.  Sigh, this is great to have people point out ideas I wouldn't have imagined.  Thank you for the awesome suggestion.  Please leave this PR open and I'll see if I can work something up here...

**Comment by pschou** [comment]
@SuperQ (and @gaissmai)  Please let me know how this looks!

**Comment by pschou** [comment]
@SuperQ or @TimRots, do you have any feedback items for me on this implementation?

**Comment by SuperQ** [comment]
Sorry, I haven't had a chance to review the updates.

**Comment by pschou** [comment]
@SuperQ, okay cool I look forward to your review

**Comment by TimRots** [comment]
> @SuperQ or @TimRots, do you have any feedback items for me on this implementation?

Hi @pschou, thank you for your patience over the holidays :)
The idea that you propose is super interesting and gives some foods for thoughts. 

How would you feel about the approach @SuperQ mentioned in his [comment](https://github.com/gosnmp/gosnmp/pull/276#issuecomment-745428435) earlier, using multiple function hooks instead of using a single function?

some pseudo:
```
    // OnSent is called when a packet is sent
    OnSent func(*GoSNMP)

    // OnRecv is called when a packet is received
    OnRecv func(*GoSNMP)

    // OnRetry is called when a retry attempt is done
    OnRetry func(*GoSNMP)

    // OnFinish is called when the request completed
    OnFinish func(*GoSNMP.Stats)
...
    @@ -137,6 +156,7 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
    for retries := 0; ; retries++ {
        if retries > 0 {
            x.OnRetry = func(stats *....) {
                x.timekeeper(Retry)
            }
            x.logPrintf("Retry number %d. Last error was: %v", retries, err)
         }
```

**Comment by pschou** [comment]
Yes, I saw that and reviewed the ping command.  I could see benefits for either way, one function with an input of a state, or many definable handles each with individual calls.  With your pseudo code of passing the current SNMP structure could help with knowing which function is being processed.  I see you had a .Stats element in the last call.  Maybe this would help with deciphering a walk instead of bulkget

**Comment by pschou** [comment]
@TimRots When you have a chance, please take a look and let me know how this looks. All the best!

**Comment by TimRots** [comment]
Thank you for your efforts @pschou ! I like this approach a lot.

I think it would be nice to have an example added in the ``examples`` dir, which demonstrates how to use the new feature.

@SuperQ what are you thinking?

File: gosnmp.go
Comment (SuperQ) [inline gosnmp.go:87]: Minor wording suggestion.

```suggestion
	// Message hook methods allow passing in a functions at various points in the packet handling.
	// For example, this can be used to collect packet timing, add metrics, or implement tracing.
	// PreSend is called before a packet is sent.
```

File: gosnmp.go
Comment (SuperQ) [inline gosnmp.go:90]: Please end all comments with a `.`

**Review by SuperQ — COMMENTED** [review]
Awesome, I like it. Some minor wording nits.

It might also be worth adding an example of usage. Maybe a Prometheus metrics example?

**Comment by pschou** [comment]
@SuperQ Thank you for the suggestions.  I am hoping to get them in place soon.

In terms of a working example / proemetheus exporter, here is a snippet from  https://github.com/pschou/snmp-prom

	var sent time.Time
	snmp.OnSent = func(x *g.GoSNMP) {
		sent = time.Now()
	}
	snmp.OnRecv = func(x *g.GoSNMP) {
		config.Devices[idev].latency = time.Since(sent).Nanoseconds()
	}
	dev_data, err := snmp.Get(oids) // Get() accepts up to g.MAX_OIDS
	snmp.OnSent = nil
	snmp.OnRecv = nil
	if err != nil {
		log.Println("Error during oid fetch for device status host:", config.Devices[idev].Host, err)
		return
	}

**Comment by pschou** [comment]
@SuperQ I finally got around to fixing these things up.  Thank you for your attention and passion to seeing this through.  I found the readme was pointing to an outdated file when I was looking up which example to use.  The example2 seems more fitting for putting an example as it isn't in the first one to avoid confusion (to keep things simple).

What visions do you have for a Prometheus exporter example?

**Comment by SuperQ** [comment]
I was thinking about something simple like

```
var sent time.Time
snmpLatency := promauto.NewHistogram(
	prometheus.HistogramOpts{
		Name: "snmp_response_duration_seconds",
		Help: "SNMP packet response latency",
		Buckets: prometheus.DefBuckets,
	},
)

snmp.OnSent = func(x *g.GoSNMP) {
	sent = time.Now()
}
snmp.OnRecv = func(x *g.GoSNMP) {
	snmpLatency.Observe(time.Since(sent).Seconds())
}
```

**Comment by pschou** [comment]
@SuperQ  When you have a moment, please take a look at examples/example_prom/main.go and see if this does what you are looking for.  I decided to give a broader view of what is possible.  Maybe too much?

**Comment by SuperQ** [comment]
The main.go example is a bit much. Maybe it's fine. I was just thinking about a small example in a comment for the godoc references.

File: examples/example_prom/main.go
Comment (SuperQ) [inline examples/example_prom/main.go:84]: It's against best practices to include location labels in metrics. These are assigned in the Prometheus config.

Really, the best practice here is to not include any labels. It's a bit noisy for this simple example code and encourages users to do the same.

File: examples/example_prom/main.go
Comment (SuperQ) [inline examples/example_prom/main.go:102]: This is unnecessary, Prometheus decides what to do.

**Comment by SuperQ** [comment]
Looking more closely over the example code, this is probably not something we want to include. It does not conform with Prometheus design best practices.

**Comment by pschou** [comment]
Thank you for the feedback and I agree with the constant labels mark, if this tool / example marks one device, that should be specified in the configuration like you said.  This is my error and overstep in thinking this.  My thought process was I usually check multiple things via one proemetheus scrape... Nope.

With regards to the timing / caching, I am open to suggestions.  The experience / problem I had with collecting snmp metrics on a well known vendor device, is that I have inadvertently broke the network device by hammering it every 15 seconds (self inflicted DoS).  So, I have since then, been making sure that this doesn't happen again by putting limiting queries on the end point which connects to the device.  When I retire, I hope that by doing this early, not only will I save myself from myself, but also who ever takes up the job after me.  I have this in the comment above, but maybe I should repeat that in this SleepDuration... or better yet call it a hold off to prevent hammering?  :)

**Comment by SuperQ** [comment]
For hammering network devices, that's what test labs are for. :grin: I know that not all devices respond well to SNMP.

The point is that I don't throw out best practices when I hit a speedbump. A self-induced DoS is a learning event for the sharp edges of a tool, not proof that the tool is inherently flawed. The "broken" class of devices will be configured with their own settings, separate from the normal workflow.

For example, I just inherited some old ex4200 switches that can't respond to ifXTable in under 60 seconds when stacked. They have a flawed SNMP implementation that is just too slow for the normal `if_mib`. So I created a separately tuned snmp_exporter module and have been tuning the on-device settings.

**Comment by pschou** [comment]
Haha, yes, test labs.  All too often I'm fixing this for others.  The "DevOps" SRE concept, is often misunderstood by developers / leaders as omitting that "integration" step and making the Dev+Ops an oxymoron by pushing to "production" too quickly.  Test every change!  Sorry for my soap box.

I did draw code & concepts from the project SuperQ/smokeping_prober which does a similar thing with out of band timing, but albeit, smokeping_prober works in a much shorter scale.  I'm at an impasse, with not knowing where to go from here; I'll defer to your best judgment on what this example should and should not do.  

Maybe make two?  I don't like the idea of putting the team on the hook for supporting / updating more code than needed, so never mind there.

**Comment by SuperQ** [comment]
The timing hook part is completely fine, I just think we don't need a full working example main.go. I was only suggesting we have a documentation example snipit, not a full piece of code.

**Comment by pschou** [comment]
Oops, sorry about that!  I'll grab a section out and call it done.

**Comment by pschou** [comment]
@SuperQ random questions here, with the ex4200, what tuning did you have to do to handle the 60 second reply delay?  Was this delay seen for the first OID get or per walk request?

**Comment by pschou** [comment]
@TimRots @SuperQ thank you both for your review and more over the support in keeping this project going!  With the main topic of the PR, timing function handles, I have added, tested, and updated example2/main.go to show these calls in action.

The idea of adding a proemetheus scraping endpoint seem like it may be outside this PR / my level of effort for this ask to document how to use these function handles.  I have built a fully functional example of a proemetheus endpoint which is not that short and simple (like an easy example).  I'm struggling here on how to capture and show the functionality without showing a full example.

Would you be so kind to accept this PR?

All the best!

**Comment by SuperQ** [comment]
Please drop `examples/example_prom/main.go`, we don't need it.

**Comment by pschou** [comment]
@SuperQ No problem-- it should be gone now, please confirm.

**Review by SuperQ — APPROVED** [review]
LGTM

**Review by TimRots — APPROVED** [review]
LGTM

**Pull request merged**
